gettext-tools/man/msgattrib.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.24.
.TH MSGATTRIB "1" "December 2015" "GNU gettext-tools 0.19.7" GNU
.SH NAME
msgattrib \- attribute matching and manipulation on message catalog
.SH SYNOPSIS
.B msgattrib
[\fIOPTION\fR] [\fIINPUTFILE\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Filters the messages of a translation catalog according to their attributes,
and manipulates the attributes.
.PP
Mandatory arguments to long options are mandatory for short options too.
.SS "Input file location:"
.TP
INPUTFILE
input PO file
.TP
\fB\-D\fR, \fB\-\-directory\fR=\fIDIRECTORY\fR
add DIRECTORY to list for input files search
.PP
If no input file is given or if it is -, standard input is read.
.SS "Output file location:"
.TP
\fB\-o\fR, \fB\-\-output\-file\fR=\fIFILE\fR
write output to specified file
.PP
The results are written to standard output if no output file is specified
or if it is -.
.SS "Message selection:"
.TP
\fB\-\-translated\fR
keep translated, remove untranslated messages
.TP
\fB\-\-untranslated\fR
keep untranslated, remove translated messages
.TP
\fB\-\-no\-fuzzy\fR
remove 'fuzzy' marked messages
.TP
\fB\-\-only\-fuzzy\fR
keep 'fuzzy' marked messages
.TP
\fB\-\-no\-obsolete\fR
remove obsolete #~ messages
.TP
\fB\-\-only\-obsolete\fR
keep obsolete #~ messages
.SS "Attribute manipulation:"
.TP
\fB\-\-set\-fuzzy\fR
set all messages 'fuzzy'
.TP
\fB\-\-clear\-fuzzy\fR
set all messages non-'fuzzy'
.TP
\fB\-\-set\-obsolete\fR
set all messages obsolete
.TP
\fB\-\-clear\-obsolete\fR
set all messages non-obsolete
.TP
\fB\-\-previous\fR
when setting 'fuzzy', keep previous msgids
of translated messages.
.TP
\fB\-\-clear\-previous\fR
remove the "previous msgid" from all messages
.TP
\fB\-\-empty\fR
when removing 'fuzzy', also set msgstr empty
.TP
\fB\-\-only\-file\fR=\fIFILE\fR.po
manipulate only entries listed in FILE.po
.TP
\fB\-\-ignore\-file\fR=\fIFILE\fR.po
manipulate only entries not listed in FILE.po
.TP
\fB\-\-fuzzy\fR
synonym for \fB\-\-only\-fuzzy\fR \fB\-\-clear\-fuzzy\fR
.TP
\fB\-\-obsolete\fR
synonym for \fB\-\-only\-obsolete\fR \fB\-\-clear\-obsolete\fR
.SS "Input file syntax:"
.TP
\fB\-P\fR, \fB\-\-properties\-input\fR
input file is in Java .properties syntax
.TP
\fB\-\-stringtable\-input\fR
input file is in NeXTstep/GNUstep .strings syntax
.SS "Output details:"
.TP
\fB\-\-color\fR
use colors and other text attributes always
.TP
\fB\-\-color\fR=\fIWHEN\fR
use colors and other text attributes if WHEN.
WHEN may be 'always', 'never', 'auto', or 'html'.
.TP
\fB\-\-style\fR=\fISTYLEFILE\fR
specify CSS style rule file for \fB\-\-color\fR
.TP
\fB\-e\fR, \fB\-\-no\-escape\fR
do not use C escapes in output (default)
.TP
\fB\-E\fR, \fB\-\-escape\fR
use C escapes in output, no extended chars
.TP
\fB\-\-force\-po\fR
write PO file even if empty
.TP
\fB\-i\fR, \fB\-\-indent\fR
write the .po file using indented style
.TP
\fB\-\-no\-location\fR
do not write '#: filename:line' lines
.TP
\fB\-n\fR, \fB\-\-add\-location\fR
generate '#: filename:line' lines (default)
.TP
\fB\-\-strict\fR
write out strict Uniforum conforming .po file
.TP
\fB\-p\fR, \fB\-\-properties\-output\fR
write out a Java .properties file
.TP
\fB\-\-stringtable\-output\fR
write out a NeXTstep/GNUstep .strings file
.TP
\fB\-w\fR, \fB\-\-width\fR=\fINUMBER\fR
set output page width
.TP
\fB\-\-no\-wrap\fR
do not break long message lines, longer than
the output page width, into several lines
.TP
\fB\-s\fR, \fB\-\-sort\-output\fR
generate sorted output
.TP
\fB\-F\fR, \fB\-\-sort\-by\-file\fR
sort output by file location
.SS "Informative output:"
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
output version information and exit
.SH AUTHOR
Written by Bruno Haible.
.SH "REPORTING BUGS"
Report bugs to <bug-gnu-gettext@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2001-2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B msgattrib
is maintained as a Texinfo manual.  If the
.B info
and
.B msgattrib
programs are properly installed at your site, the command
.IP
.B info msgattrib
.PP
should give you access to the complete manual.


NEWS
Version 0.19.7 - December 2015

* Programming languages support:
  - XML:
    xgettext can now load custom string extraction rules supplied by
    consumer projects.  The rules are written in XML, utilizing the
    Internationalization Tag Set (ITS) standard.  All the existing
    XML-based language scanners (Glade, GSettings, and AppData) are
    rewritten using ITS.  In addition, msgfmt now has --xml option to
    merge translations back to the original XML document.

* Portability:
  - Improve OS/2 kLIBC support (still not complete)
  - Remove dependency on expat

Version 0.19.6 - September 2015

* Programming languages support:
  - AppData:
    xgettext now supports AppData file format, used by software center
    applications (e.g., GNOME Software) to describe installable
    applications.

* A new macro AM_GNU_GETTEXT_REQUIRE_VERSION can be used to indicate
  autopoint to pull the latest available infrastructure, instead of
  the exact version specified with AM_GNU_GETTEXT_VERSION.  When
  AM_GNU_GETTEXT_REQUIRE_VERSION is used, AM_GNU_GETTEXT_VERSION is
  ignored.

* po/Makefile.in.in can now insert the file $(DOMAIN).pot-header to
  $(DOMAIN).pot, instead of the standard header comments.

* Bug fixes:
  - Fix mishandling of gettext version numbers for minor releases, in
    po-mode.el and gettextize.
  - Fix build with --enable-relocatable.

Version 0.19.5 - July 2015

* xgettext now has a feature to perform syntax checks on msgid, which
  could enforce common styles of translatable strings, such as to
  prefer Unicode characters to the corresponding ASCII characters.
  They can be enabled with --check option or special "xgettext: "
  comment in the source code.  By default, no syntax checks are
  enabled.

* msgfilter and msgexec now have an option --newline, which appends a
  newline character to filter input and trims it from the filter
  output.  This would allow filter programs to be more POSIX friendly.

* The base Unicode standard is now updated to 8.0.0.  This
  particularly improves "\N{...}" notation handling of xgettext for
  Perl and Python.

* msginit is now capable of generating "Plural-Forms:" from Unicode
  CLDR.  This feature is still experimental, but you can try it by
  setting the GETTEXTCLDRDIR environment variable pointing to the
  directory where the CLDR archive is extracted.  The actual
  conversion is done by a helper program 'cldr-plural', which can be
  used as a generic converter and evaluator of CLDR plural forms.

* Programming languages support:
  - C++ with KDE: xgettext and msgfmt can now recognize KUIT (KDE User
    Interface Text) markup.  See the documentation section "KUIT
    Format Strings" for more info.
  - C++ with KDE: xgettext now recognizes all default KDE keywords.
    This removes the need for a long list of --keyword and --flag
    options to perform a reasonable extraction.

* Bug fixes:
  - xgettext C++11 raw string recognition is now stricter and don't
    accept unbalanced delimiters.
  - Suppress baseless warnings which msgfmt emits when processing a
    .desktop file.
  - xgettext line wrapping behaviour is now consistent between comment
    lines and non-comment lines.
  - Fix msgfilter-7 test failure on some platforms.
  - Fix VPATH build.

Version 0.19.4 - December 2014

* The --keyword option of xgettext now accepts same argument number
  for both singular and plural forms.

* Programming languages support:
  - C#: xgettext now properly handles Unicode characters encoded with
    surrogate pairs.
  - C/C++: xgettext now recognizes ISO/IEC 9899:2011 string literals
    prefixed by R, u8, u8R, u, uR, U, UR, L, or LR.
  - Shell: xgettext now properly recognizes Bash ANSI-C quoting ($'...').

* Bug fixes:
  - Fix integer overflow when reading certain MO files with msgunfmt.
  - Avoid invalid memory access in various cases.  In particular, when
    the same argument number is specified for singular/plural
    arguments, and when checking Lisp and Scheme format strings.

* Portability:
  - Building on Mac OS X 10.10 and AIX 7.1 is now supported.

Version 0.19.3 - October 2014

* Bug fixes:
  - Fix xgettext mishandling of octal character escapes in C.
  - Fix autopoint infinite recursion with certain configure.ac.

* The po/Makevars file has a new field MSGINIT_OPTIONS, that can be
  used to adjust msginit's operation.  This is particularly useful for
  controlling line wrapping behavior together with MSGMERGE_OPTIONS
  and XGETTEXT_OPTIONS.

* Portability:
  - Building on Solaris 10 and 11 with Solaris Studio compiler is now
    fixed.

Version 0.19.2 - July 2014

* Bug fixes:
  - Fix xgettext crash in parsing empty string literals in C and Vala.
  - ChangeLog file is added back to the gettext infrastructure.  It was
    mistakenly removed in 0.19.
  - Autoconf macro trace in autopoint now works again with Autoconf 2.68
    or earlier.  It was a regression in 0.19.

Version 0.19.1 - June 2014

* Programming languages support:

  - Desktop Entry:
    msgfmt now always reads the po/LINGUAS file, regardless of whether
    the LINGUAS environment variable is set.  The variable can now be
    used to restrict the languages list read from the po/LINGUAS file.

  - Vala:
    Bug fix in xgettext handling of "//" in string literals.  This was
    a regression after the C-99 Unicode escape support.

* The po/Makevars.template file now contains the newly added variables.


gettext-tools/src/msgattrib.c
/* Manipulates attributes of messages in translation catalogs.
   Copyright (C) 2001-2007, 2009-2010, 2012 Free Software Foundation, Inc.
   Written by Bruno Haible <haible@clisp.cons.org>, 2001.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <getopt.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <locale.h>

#include "closeout.h"
#include "dir-list.h"
#include "error.h"
#include "error-progname.h"
#include "progname.h"
#include "relocatable.h"
#include "basename.h"
#include "message.h"
#include "read-catalog.h"
#include "read-po.h"
#include "read-properties.h"
#include "read-stringtable.h"
#include "write-catalog.h"
#include "write-po.h"
#include "write-properties.h"
#include "write-stringtable.h"
#include "color.h"
#include "propername.h"
#include "gettext.h"

#define _(str) gettext (str)


/* Force output of PO file even if empty.  */
static int force_po;

/* Bit mask of subsets to remove.  */
enum
{
  REMOVE_UNTRANSLATED   = 1 << 0,
  REMOVE_TRANSLATED     = 1 << 1,
  REMOVE_FUZZY          = 1 << 2,
  REMOVE_NONFUZZY       = 1 << 3,
  REMOVE_OBSOLETE       = 1 << 4,
  REMOVE_NONOBSOLETE    = 1 << 5
};
static int to_remove;

/* Bit mask of actions to perform on all messages.  */
enum
{
  SET_FUZZY             = 1 << 0,
  RESET_FUZZY           = 1 << 1,
  SET_OBSOLETE          = 1 << 2,
  RESET_OBSOLETE        = 1 << 3,
  REMOVE_PREV           = 1 << 4
};
static int to_change;

/* Long options.  */
static const struct option long_options[] =
{
  { "add-location", no_argument, &line_comment, 1 },
  { "clear-fuzzy", no_argument, NULL, CHAR_MAX + 8 },
  { "clear-obsolete", no_argument, NULL, CHAR_MAX + 10 },
  { "clear-previous", no_argument, NULL, CHAR_MAX + 18 },
  { "color", optional_argument, NULL, CHAR_MAX + 19 },
  { "directory", required_argument, NULL, 'D' },
  { "escape", no_argument, NULL, 'E' },
  { "force-po", no_argument, &force_po, 1 },
  { "fuzzy", no_argument, NULL, CHAR_MAX + 11 },
  { "help", no_argument, NULL, 'h' },
  { "ignore-file", required_argument, NULL, CHAR_MAX + 15 },
  { "indent", no_argument, NULL, 'i' },
  { "no-escape", no_argument, NULL, 'e' },
  { "no-fuzzy", no_argument, NULL, CHAR_MAX + 3 },
  { "no-location", no_argument, &line_comment, 0 },
  { "no-obsolete", no_argument, NULL, CHAR_MAX + 5 },
  { "no-wrap", no_argument, NULL, CHAR_MAX + 13 },
  { "obsolete", no_argument, NULL, CHAR_MAX + 12 },
  { "only-file", required_argument, NULL, CHAR_MAX + 14 },
  { "only-fuzzy", no_argument, NULL, CHAR_MAX + 4 },
  { "only-obsolete", no_argument, NULL, CHAR_MAX + 6 },
  { "output-file", required_argument, NULL, 'o' },
  { "properties-input", no_argument, NULL, 'P' },
  { "properties-output", no_argument, NULL, 'p' },
  { "set-fuzzy", no_argument, NULL, CHAR_MAX + 7 },
  { "set-obsolete", no_argument, NULL, CHAR_MAX + 9 },
  { "sort-by-file", no_argument, NULL, 'F' },
  { "sort-output", no_argument, NULL, 's' },
  { "stringtable-input", no_argument, NULL, CHAR_MAX + 16 },
  { "stringtable-output", no_argument, NULL, CHAR_MAX + 17 },
  { "strict", no_argument, NULL, 'S' },
  { "style", required_argument, NULL, CHAR_MAX + 20 },
  { "translated", no_argument, NULL, CHAR_MAX + 1 },
  { "untranslated", no_argument, NULL, CHAR_MAX + 2 },
  { "version", no_argument, NULL, 'V' },
  { "width", required_argument, NULL, 'w', },
  { NULL, 0, NULL, 0 }
};


/* Forward declaration of local functions.  */
static void usage (int status)
#if defined __GNUC__ && ((__GNUC__ == 2 && __GNUC_MINOR__ >= 5) || __GNUC__ > 2)
        __attribute__ ((noreturn))
#endif
;
static msgdomain_list_ty *process_msgdomain_list (msgdomain_list_ty *mdlp,
                                                  msgdomain_list_ty *only_mdlp,
                                                msgdomain_list_ty *ignore_mdlp);


int
main (int argc, char **argv)
{
  int optchar;
  bool do_help;
  bool do_version;
  char *output_file;
  const char *input_file;
  const char *only_file;
  const char *ignore_file;
  msgdomain_list_ty *only_mdlp;
  msgdomain_list_ty *ignore_mdlp;
  msgdomain_list_ty *result;
  catalog_input_format_ty input_syntax = &input_format_po;
  catalog_output_format_ty output_syntax = &output_format_po;
  bool sort_by_msgid = false;
  bool sort_by_filepos = false;

  /* Set program name for messages.  */
  set_program_name (argv[0]);
  error_print_progname = maybe_print_progname;

#ifdef HAVE_SETLOCALE
  /* Set locale via LC_ALL.  */
  setlocale (LC_ALL, "");
#endif

  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, relocate (LOCALEDIR));
  bindtextdomain ("bison-runtime", relocate (BISON_LOCALEDIR));
  textdomain (PACKAGE);

  /* Ensure that write errors on stdout are detected.  */
  atexit (close_stdout);

  /* Set default values for variables.  */
  do_help = false;
  do_version = false;
  output_file = NULL;
  input_file = NULL;
  only_file = NULL;
  ignore_file = NULL;

  while ((optchar = getopt_long (argc, argv, "D:eEFhino:pPsVw:", long_options,
                                 NULL)) != EOF)
    switch (optchar)
      {
      case '\0':                /* Long option.  */
        break;

      case 'D':
        dir_list_append (optarg);
        break;

      case 'e':
        message_print_style_escape (false);
        break;

      case 'E':
        message_print_style_escape (true);
        break;

      case 'F':
        sort_by_filepos = true;
        break;

      case 'h':
        do_help = true;
        break;

      case 'i':
        message_print_style_indent ();
        break;

      case 'n':
        line_comment = 1;
        break;

      case 'o':
        output_file = optarg;
        break;

      case 'p':
        output_syntax = &output_format_properties;
        break;

      case 'P':
        input_syntax = &input_format_properties;
        break;

      case 's':
        sort_by_msgid = true;
        break;

      case 'S':
        message_print_style_uniforum ();
        break;

      case 'V':
        do_version = true;
        break;

      case 'w':
        {
          int value;
          char *endp;
          value = strtol (optarg, &endp, 10);
          if (endp != optarg)
            message_page_width_set (value);
        }
        break;

      case CHAR_MAX + 1: /* --translated */
        to_remove |= REMOVE_UNTRANSLATED;
        break;

      case CHAR_MAX + 2: /* --untranslated */
        to_remove |= REMOVE_TRANSLATED;
        break;

      case CHAR_MAX + 3: /* --no-fuzzy */
        to_remove |= REMOVE_FUZZY;
        break;

      case CHAR_MAX + 4: /* --only-fuzzy */
        to_remove |= REMOVE_NONFUZZY;
        break;

      case CHAR_MAX + 5: /* --no-obsolete */
        to_remove |= REMOVE_OBSOLETE;
        break;

      case CHAR_MAX + 6: /* --only-obsolete */
        to_remove |= REMOVE_NONOBSOLETE;
        break;

      case CHAR_MAX + 7: /* --set-fuzzy */
        to_change |= SET_FUZZY;
        break;

      case CHAR_MAX + 8: /* --clear-fuzzy */
        to_change |= RESET_FUZZY;
        break;

      case CHAR_MAX + 9: /* --set-obsolete */
        to_change |= SET_OBSOLETE;
        break;

      case CHAR_MAX + 10: /* --clear-obsolete */
        to_change |= RESET_OBSOLETE;
        break;

      case CHAR_MAX + 11: /* --fuzzy */
        to_remove |= REMOVE_NONFUZZY;
        to_change |= RESET_FUZZY;
        break;

      case CHAR_MAX + 12: /* --obsolete */
        to_remove |= REMOVE_NONOBSOLETE;
        to_change |= RESET_OBSOLETE;
        break;

      case CHAR_MAX + 13: /* --no-wrap */
        message_page_width_ignore ();
        break;

      case CHAR_MAX + 14: /* --only-file */
        only_file = optarg;
        break;

      case CHAR_MAX + 15: /* --ignore-file */
        ignore_file = optarg;
        break;

      case CHAR_MAX + 16: /* --stringtable-input */
        input_syntax = &input_format_stringtable;
        break;

      case CHAR_MAX + 17: /* --stringtable-output */
        output_syntax = &output_format_stringtable;
        break;

      case CHAR_MAX + 18: /* --clear-previous */
        to_change |= REMOVE_PREV;
        break;

      case CHAR_MAX + 19: /* --color */
        if (handle_color_option (optarg) || color_test_mode)
          usage (EXIT_FAILURE);
        break;

      case CHAR_MAX + 20: /* --style */
        handle_style_option (optarg);
        break;

      default:
        usage (EXIT_FAILURE);
        /* NOTREACHED */
      }

  /* Version information requested.  */
  if (do_version)
    {
      printf ("%s (GNU %s) %s\n", basename (program_name), PACKAGE, VERSION);
      /* xgettext: no-wrap */
      printf (_("Copyright (C) %s Free Software Foundation, Inc.\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n\
"),
              "2001-2010");
      printf (_("Written by %s.\n"), proper_name ("Bruno Haible"));
      exit (EXIT_SUCCESS);
    }

  /* Help is requested.  */
  if (do_help)
    usage (EXIT_SUCCESS);

  /* Test whether we have an .po file name as argument.  */
  if (optind == argc)
    input_file = "-";
  else if (optind + 1 == argc)
    input_file = argv[optind];
  else
    {
      error (EXIT_SUCCESS, 0, _("at most one input file allowed"));
      usage (EXIT_FAILURE);
    }

  /* Verify selected options.  */
  if (!line_comment && sort_by_filepos)
    error (EXIT_FAILURE, 0, _("%s and %s are mutually exclusive"),
           "--no-location", "--sort-by-file");

  if (sort_by_msgid && sort_by_filepos)
    error (EXIT_FAILURE, 0, _("%s and %s are mutually exclusive"),
           "--sort-output", "--sort-by-file");

  /* Read input file.  */
  result = read_catalog_file (input_file, input_syntax);

  /* Read optional files that limit the extent of the attribute changes.  */
  only_mdlp = (only_file != NULL
               ? read_catalog_file (only_file, input_syntax)
               : NULL);
  ignore_mdlp = (ignore_file != NULL
                 ? read_catalog_file (ignore_file, input_syntax)
                 : NULL);

  /* Filter the messages and manipulate the attributes.  */
  result = process_msgdomain_list (result, only_mdlp, ignore_mdlp);

  /* Sorting the list of messages.  */
  if (sort_by_filepos)
    msgdomain_list_sort_by_filepos (result);
  else if (sort_by_msgid)
    msgdomain_list_sort_by_msgid (result);

  /* Write the PO file.  */
  msgdomain_list_print (result, output_file, output_syntax, force_po, false);

  exit (EXIT_SUCCESS);
}


/* Display usage information and exit.  */
static void
usage (int status)
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, _("Try '%s --help' for more information.\n"),
             program_name);
  else
    {
      printf (_("\
Usage: %s [OPTION] [INPUTFILE]\n\
"), program_name);
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Filters the messages of a translation catalog according to their attributes,\n\
and manipulates the attributes.\n"));
      printf ("\n");
      printf (_("\
Mandatory arguments to long options are mandatory for short options too.\n"));
      printf ("\n");
      printf (_("\
Input file location:\n"));
      printf (_("\
  INPUTFILE                   input PO file\n"));
      printf (_("\
  -D, --directory=DIRECTORY   add DIRECTORY to list for input files search\n"));
      printf (_("\
If no input file is given or if it is -, standard input is read.\n"));
      printf ("\n");
      printf (_("\
Output file location:\n"));
      printf (_("\
  -o, --output-file=FILE      write output to specified file\n"));
      printf (_("\
The results are written to standard output if no output file is specified\n\
or if it is -.\n"));
      printf ("\n");
      printf (_("\
Message selection:\n"));
      printf (_("\
      --translated            keep translated, remove untranslated messages\n"));
      printf (_("\
      --untranslated          keep untranslated, remove translated messages\n"));
      printf (_("\
      --no-fuzzy              remove 'fuzzy' marked messages\n"));
      printf (_("\
      --only-fuzzy            keep 'fuzzy' marked messages\n"));
      printf (_("\
      --no-obsolete           remove obsolete #~ messages\n"));
      printf (_("\
      --only-obsolete         keep obsolete #~ messages\n"));
      printf ("\n");
      printf (_("\
Attribute manipulation:\n"));
      printf (_("\
      --set-fuzzy             set all messages 'fuzzy'\n"));
      printf (_("\
      --clear-fuzzy           set all messages non-'fuzzy'\n"));
      printf (_("\
      --set-obsolete          set all messages obsolete\n"));
      printf (_("\
      --clear-obsolete        set all messages non-obsolete\n"));
      printf (_("\
      --clear-previous        remove the \"previous msgid\" from all messages\n"));
      printf (_("\
      --only-file=FILE.po     manipulate only entries listed in FILE.po\n"));
      printf (_("\
      --ignore-file=FILE.po   manipulate only entries not listed in FILE.po\n"));
      printf (_("\
      --fuzzy                 synonym for --only-fuzzy --clear-fuzzy\n"));
      printf (_("\
      --obsolete              synonym for --only-obsolete --clear-obsolete\n"));
      printf ("\n");
      printf (_("\
Input file syntax:\n"));
      printf (_("\
  -P, --properties-input      input file is in Java .properties syntax\n"));
      printf (_("\
      --stringtable-input     input file is in NeXTstep/GNUstep .strings syntax\n"));
      printf ("\n");
      printf (_("\
Output details:\n"));
      printf (_("\
      --color                 use colors and other text attributes always\n\
      --color=WHEN            use colors and other text attributes if WHEN.\n\
                              WHEN may be 'always', 'never', 'auto', or 'html'.\n"));
      printf (_("\
      --style=STYLEFILE       specify CSS style rule file for --color\n"));
      printf (_("\
  -e, --no-escape             do not use C escapes in output (default)\n"));
      printf (_("\
  -E, --escape                use C escapes in output, no extended chars\n"));
      printf (_("\
      --force-po              write PO file even if empty\n"));
      printf (_("\
  -i, --indent                write the .po file using indented style\n"));
      printf (_("\
      --no-location           do not write '#: filename:line' lines\n"));
      printf (_("\
  -n, --add-location          generate '#: filename:line' lines (default)\n"));
      printf (_("\
      --strict                write out strict Uniforum conforming .po file\n"));
      printf (_("\
  -p, --properties-output     write out a Java .properties file\n"));
      printf (_("\
      --stringtable-output    write out a NeXTstep/GNUstep .strings file\n"));
      printf (_("\
  -w, --width=NUMBER          set output page width\n"));
      printf (_("\
      --no-wrap               do not break long message lines, longer than\n\
                              the output page width, into several lines\n"));
      printf (_("\
  -s, --sort-output           generate sorted output\n"));
      printf (_("\
  -F, --sort-by-file          sort output by file location\n"));
      printf ("\n");
      printf (_("\
Informative output:\n"));
      printf (_("\
  -h, --help                  display this help and exit\n"));
      printf (_("\
  -V, --version               output version information and exit\n"));
      printf ("\n");
      /* TRANSLATORS: The placeholder indicates the bug-reporting address
         for this package.  Please add _another line_ saying
         "Report translation bugs to <...>\n" with the address for translation
         bugs (typically your translation team's web or email address).  */
      fputs (_("Report bugs to <bug-gnu-gettext@gnu.org>.\n"),
             stdout);
    }

  exit (status);
}


/* Return true if a message should be kept.  */
static bool
is_message_selected (const message_ty *mp)
{
  /* Always keep the header entry.  */
  if (is_header (mp))
    return true;

  if ((to_remove & (REMOVE_UNTRANSLATED | REMOVE_TRANSLATED))
      && (mp->msgstr[0] == '\0'
          ? to_remove & REMOVE_UNTRANSLATED
          : to_remove & REMOVE_TRANSLATED))
    return false;

  if ((to_remove & (REMOVE_FUZZY | REMOVE_NONFUZZY))
      && (mp->is_fuzzy
          ? to_remove & REMOVE_FUZZY
          : to_remove & REMOVE_NONFUZZY))
    return false;

  if ((to_remove & (REMOVE_OBSOLETE | REMOVE_NONOBSOLETE))
      && (mp->obsolete
          ? to_remove & REMOVE_OBSOLETE
          : to_remove & REMOVE_NONOBSOLETE))
    return false;

  return true;
}


static void
process_message_list (message_list_ty *mlp,
                      message_list_ty *only_mlp, message_list_ty *ignore_mlp)
{
  /* Keep only the selected messages.  */
  message_list_remove_if_not (mlp, is_message_selected);

  /* Change the attributes.  */
  if (to_change)
    {
      size_t j;

      for (j = 0; j < mlp->nitems; j++)
        {
          message_ty *mp = mlp->item[j];

          /* Attribute changes only affect messages listed in --only-file
             and not listed in --ignore-file.  */
          if ((only_mlp
               ? message_list_search (only_mlp, mp->msgctxt, mp->msgid) != NULL
               : true)
              && (ignore_mlp
                  ? message_list_search (ignore_mlp, mp->msgctxt, mp->msgid) == NULL
                  : true))
            {
              if (to_change & SET_FUZZY)
                mp->is_fuzzy = true;
              if (to_change & RESET_FUZZY)
                mp->is_fuzzy = false;
              /* Always keep the header entry non-obsolete.  */
              if ((to_change & SET_OBSOLETE) && !is_header (mp))
                mp->obsolete = true;
              if (to_change & RESET_OBSOLETE)
                mp->obsolete = false;
              if (to_change & REMOVE_PREV)
                {
                  mp->prev_msgctxt = NULL;
                  mp->prev_msgid = NULL;
                  mp->prev_msgid_plural = NULL;
                }
            }
        }
    }
}


static msgdomain_list_ty *
process_msgdomain_list (msgdomain_list_ty *mdlp,
                        msgdomain_list_ty *only_mdlp,
                        msgdomain_list_ty *ignore_mdlp)
{
  size_t k;

  for (k = 0; k < mdlp->nitems; k++)
    process_message_list (mdlp->item[k]->messages,
                          only_mdlp
                          ? msgdomain_list_sublist (only_mdlp,
                                                    mdlp->item[k]->domain,
                                                    true)
                          : NULL,
                          ignore_mdlp
                          ? msgdomain_list_sublist (ignore_mdlp,
                                                    mdlp->item[k]->domain,
                                                    false)
                          : NULL);

  return mdlp;
}
